### PR TITLE
plugin func_prefix for multiple plugin load

### PR DESCRIPTION
**Pull request by mashiike**
for https://github.com/kayac/ecspresso/issues/396

When multiple tfstates are loaded, function names collide and are overwritten, resulting in only the last one being valid.

In order to resolve the above behavior without worrying about the behavior in the Plugin, a mechanism is needed that allows function names to be registered as aliases.

Therefore, we added a prefix to the function name to be registered.

The actual use would be as follows:

```yaml
# ecspresso.yml
...
plugins:
   - name: tfstate
    config:
      url: s3://tfstate1
    func_prefix: first_
  - name: tfstate
    config:
      url: s3://tfstate2
    func_prefix: second_
```

```json
[ 
   "{{ first_tfstate `aws_s3_bucket.main.arn` }}",
   "{{ second_tfstate `aws_s3_bucket.main.arn` }}"
]
```

### CODE REVIEW

File: plugin.go [hunk 1]
```diff
@@ -26,6 +65,14 @@ func (p ConfigPlugin) Setup(c *Config) error {
 	}
 }
 
+func (p ConfigPlugin) ModifyFuncMap(funcMap template.FuncMap) template.FuncMap {
+	modified := make(template.FuncMap, len(funcMap))
+	for funcName, f := range funcMap {
+		modified[strings.ToLower(p.FuncPrefix)+funcName] = f
```
Comment (fujiwara) [inline plugin.go]: `strings.ToLower` seems unnecessary because template function names need not necessarily be in lowercase.

File: plugin.go [hunk 2]
```diff
@@ -50,6 +97,7 @@ func setupPluginTFState(p ConfigPlugin, c *Config) error {
 	if err != nil {
 		return err
 	}
+	funcs = p.ModifyFuncMap(funcs)
 	c.templateFuncs = append(c.templateFuncs, funcs)
```
Comment (fujiwara) [inline plugin.go]: Conflicts of template functions must raise errors. So we can check conflicts at this line as below.

Then we will not need to add `ConfigPlugins.check()`.

```go
for name, _ := range funcs {
    if _, exists := c.templateFuncs[name]; exists {
        return fmt.Errorf("template function %s already exists. set func_prefix to %s plugin", name, p.Name)
    }
}
```

Comment (mashiike) [reply plugin.go]: templateFuncs is of type []template.Funcs, so checking here would complicate the code.

Therefore, we will check it in ConfigPlugins.check() and return an error if we find a conflict.

Comment (mashiike) [reply plugin.go]: After much thought, I decided to put it all together in a function called AppendFuncMap.

https://github.com/kayac/ecspresso/pull/398/commits/435a1c9f603368328d04db74557123cb42bfa762

**Comment by fujiwara** [comment]
Thank you!

**Pull request merged**